### PR TITLE
[FIRRTL] Implement out-of-bounds as array extension

### PR DESCRIPTION
**Pull request by darthscsi**
Improve critical path through muxes and array indexes by not adding an out-of-bounds mux, but rather extend the arrays to power of two.

### CODE REVIEW

**Comment by darthscsi** [comment]
This is based on PPA studies.

Ideally, we would tail-extend instead of element-0-extend, but chisel users have dependencies on this behavior.

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 1]
```diff
@@ -3423,6 +3423,20 @@ Value FIRRTLLowering::createArrayIndexing(Value array, Value index) {
   auto size = hw::type_cast<hw::ArrayType>(array.getType()).getSize();
   auto valWire = builder.create<sv::WireOp>(
       hw::type_cast<hw::ArrayType>(array.getType()).getElementType());
+
+  // Extend to power of 2
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToHW/LowerToHW.cpp:3427]: Can the PR expand this comment to indicate why we're doing this? The PR may want to include the verbose `TODO` as well. It's not clear that we want this to be the final say here.

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 2]
```diff
@@ -3423,6 +3423,20 @@ Value FIRRTLLowering::createArrayIndexing(Value array, Value index) {
...
+  if (!llvm::isPowerOf2_64(size)) {
+    auto extElem =
+        builder.create<hw::ConstantOp>(APInt(llvm::Log2_64_Ceil(size), 0));
+    auto extValue = builder.create<hw::ArrayGetOp>(array, extElem);
+    SmallVector<Value> temp(llvm::NextPowerOf2(size) - size, extValue);
+    auto ext = builder.create<hw::ArrayCreateOp>(temp);
+    SmallVector<Value> temp2;
+    temp2.push_back(ext.getResult());
+    temp2.push_back(array);
+    array = builder.create<hw::ArrayConcatOp>(temp2);
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToHW/LowerToHW.cpp:3437]: Golf nit: Does this work?

```suggestion
    array = builder.create<hw::ArrayConcatOp>({ext.getResult(), array});
```

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 3]
```diff
@@ -3423,6 +3423,20 @@ Value FIRRTLLowering::createArrayIndexing(Value array, Value index) {
   auto size = hw::type_cast<hw::ArrayType>(array.getType()).getSize();
   auto valWire = builder.create<sv::WireOp>(
       hw::type_cast<hw::ArrayType>(array.getType()).getElementType());
+
+  // Extend to power of 2
+  if (!llvm::isPowerOf2_64(size)) {
+    auto extElem =
+        builder.create<hw::ConstantOp>(APInt(llvm::Log2_64_Ceil(size), 0));
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToHW/LowerToHW.cpp:3430]: This should be able to consult the constant pool:

```suggestion
    auto extElem = getOrCreateIntConstant(APInt(llvm::Log2_64_Ceil(size), 0));
```

**Review by seldridge — APPROVED** [review]
LGTM

Improving PPA while still keeping SFC-exact behavior is good.

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp (hunk 2 above)
Comment (darthscsi) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:3437]: no, at least when I tried earlier.

**Comment by youngar** [comment]
Maybe we should add a lowering option for this, just in case we find that this causes issues on some tools

**Comment by darthscsi** [comment]
I'm not too worried about tools, since this is strictly legal.  I am a bit worried about coverage, but that shouldn't stop PPA improvements.

We do need an option to not emit the mux mapping attributes all the time, since that isn't always correct.

**Comment by uenoku** [comment]
I'm wondering we can achieve the same PPA result by extending the array with `comb.replicate`. (i.e. `bitcast(replicate(a[0], n))`), which might be more readable for users but it's not clear that tools are smart enough to recognize that. 

> We do need an option to not emit the mux mapping attributes all the time, since that isn't always correct.

I'll do that in https://github.com/llvm/circt/pull/3913

**Comment by darthscsi** [comment]
I generally would prefer to keep array ops as array ops not as integers
with bitcasts.  It makes life easier to not worry about endianness, array
order, etc.

**Pull request merged**